src/sat/sat_clause.cpp

```diff
@@ -123,48 +123,80 @@ namespace sat {
 
     clause_allocator::clause_allocator():
         m_allocator("clause-allocator") {
-#ifdef _AMD64_
+#if defined(_AMD64_) 
         m_num_segments = 0;
+#if defined(Z3DEBUG)
+        m_overflow_valid = false;
+#endif
 #endif
     }
 
     clause * clause_allocator::get_clause(clause_offset cls_off) const {
-#ifdef _AMD64_
+#if defined(_AMD64_) 
+#if defined (Z3DEBUG)
+        clause const* result;
+        if (m_overflow_valid && m_cls_offset2ptr.find(cls_off, result)) {
+            return const_cast<clause*>(result);
+        }
+#endif
         return reinterpret_cast<clause *>(m_segments[cls_off & c_aligment_mask] + (static_cast<size_t>(cls_off) & ~c_aligment_mask));
 #else
         return reinterpret_cast<clause *>(cls_off);
 #endif
     }
 
-#ifdef _AMD64_
-    unsigned clause_allocator::get_segment(size_t ptr) {
+#if defined(_AMD64_) 
+    unsigned clause_allocator::get_segment(clause const* cls) {
+        size_t ptr = reinterpret_cast<size_t>(cls);
+
         SASSERT((ptr & c_aligment_mask) == 0);
         ptr &= ~0xFFFFFFFFull; // Keep only high part
         unsigned i = 0;
         for (i = 0; i < m_num_segments; ++i)
             if (m_segments[i] == ptr)
                 return i;
         i = m_num_segments;
-        m_num_segments++;
-        SASSERT(m_num_segments <= c_max_segments);
-        if (i >= c_max_segments)
+#if defined(Z3DEBUG)
+        SASSERT(i < c_max_segments);
+        if (i + 1 == c_max_segments) {
+            m_overflow_valid = true;
+            i += c_max_segments * m_cls_offset2ptr.size();
+            m_ptr2cls_offset.insert(ptr, i);
```

Comment:
Also, what is the reason of having `m_ptr2cls_offset` at all if, after we run out of normal segments, each time `clause_allocator::get_segment()` is called a new index is created at line 163 and the only lookup in `clause_allocator::get_offset()` seems to be unnecessary?